Add TensorFlow implementation of  EfficientFormer

D-Roberts opened the pull request:
# What does this PR do?

* Adding EfficientFormer computer vision model TensorFlow port (not a llm port). 
* Fixes some minor typos and a couple of differences in the PyTorch model code: 1) the non-dict / tuple return was not returning last hidden state but the state before last stage. The dict and tuple return of the encoder should be equivalent, as seen in other [models](https://github.com/huggingface/transformers/blob/main/src/transformers/models/poolformer/modeling_poolformer.py#L258). 2) Two layernorms were not using the config eps (assuming that the config is the ground truth). Let me know how you think about this.


Ran tests (CPU-only, all pass) with:
`NVIDIA_TF32_OVERRIDE=1 RUN_SLOW=1 RUN_PT_TF_CROSS_TESTS=1 py.test -vv -rA tests/models/efficientformer/test_modeling_tf_efficientformer.py `

Double checked pt and tf architecture codes with the "[EfficientFormer: Vision Transformers at MobileNet Speed](https://proceedings.neurips.cc/paper_files/paper/2022/hash/5452ad8ee6ea6e7dc41db1cbd31ba0b8-Abstract-Conference.html)" paper.

Verified on example image shapes and diffs in hidden states:
```
from transformers import EfficientFormerImageProcessor
from src.transformers.models.efficientformer.modeling_tf_efficientformer import TFEfficientFormerModel
from src.transformers.models.efficientformer.modeling_efficientformer import EfficientFormerModel 

model_tf = TFEfficientFormerModel.from_pretrained("snap-research/efficientformer-l1-300", from_pt=True)
model_pt = EfficientFormerModel.from_pretrained("snap-research/efficientformer-l1-300")

image = Image.open("./tests/fixtures/tests_samples/COCO/000000039769.png")
proc =  EfficientFormerImageProcessor.from_pretrained("snap-research/efficientformer-l1-300")
inputstf = proc(images=image, return_tensors="tf")
inputspt = proc(images=image, return_tensors="pt")
       
outtf = model_tf(**inputstf, output_hidden_states=True, training=False)
with torch.no_grad():
    outpt = model_pt(**inputspt, output_hidden_states=True)

max_diff = np.amax(np.abs(outtf[0].numpy() - outpt[0].numpy()))
print(f"last hidden diff shape: {outtf[0].shape}, last hidden diff: {max_diff}, last hidden <= 1e-4, {max_diff <= 1e-4}")
for i in range(7):
    max_diff = np.amax(np.abs(outtf[1][i].numpy() - outpt[1][i].numpy()))
    print(f"hidden state {i} shape: {outtf[1][i].shape}, diff: {max_diff}, max_diff <= 1e-4: {max_diff <= 1e-4}")
```
which gives:
```
last hidden diff shape: (1, 49, 448), last hidden diff: 2.1457672119140625e-05, last hidden <= 1e-4, True
hidden state 0 shape: (1, 48, 56, 56), diff: 7.271766662597656e-06, max_diff <= 1e-4: True
hidden state 1 shape: (1, 48, 56, 56), diff: 5.054473876953125e-05, max_diff <= 1e-4: True
hidden state 2 shape: (1, 96, 28, 28), diff: 2.9087066650390625e-05, max_diff <= 1e-4: True
hidden state 3 shape: (1, 96, 28, 28), diff: 2.3603439331054688e-05, max_diff <= 1e-4: True
hidden state 4 shape: (1, 224, 14, 14), diff: 1.6689300537109375e-05, max_diff <= 1e-4: True
hidden state 5 shape: (1, 224, 14, 14), diff: 4.1961669921875e-05, max_diff <= 1e-4: True
hidden state 6 shape: (1, 448, 7, 7), diff: 1.9550323486328125e-05, max_diff <= 1e-4: True
```

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->



## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @sgugger

Integrations:

- deepspeed: HF Trainer: @stas00, Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam

Documentation: @sgugger, @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: @sgugger
- TensorFlow: @Rocketknight1

 -->

D-Roberts committed:
Add tf code for efficientformer

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_22620). All of your documentation changes will be reflected on that endpoint.

sgugger commented:
cc @Rocketknight1

D-Roberts commented:
I am delayed a bit - will get back to it soon.

Rocketknight1 commented:
Hi @D-Roberts, just letting you know the TF team at Hugging Face is aware of this and definitely interested in the port! Please ping me or @gante whenever it's ready for review, or if you run into any issues while porting.

D-Roberts commented:
@Rocketknight1 Hi! Working on tests- hopefully will ping in a few days for review! BTW- there is an Eff[icientformer v2 t](https://github.com/snap-research/EfficientFormer/tree/main/models)hat I am interested in to try out in my applications - any plans over at HF to have that implemented?

D-Roberts commented:
hi @Rocketknight1 - do you guys have a custom implementation of the AvgPool2D Keras layers which allows for `'count_include_pad=False`' as in [torch](https://pytorch.org/docs/stable/generated/torch.nn.AvgPool2d.html)? Without this option there are diffs in model outputs and I was wondering if you ran into this before.

D-Roberts committed:
Fix return dict bug - return last hidden state after last stage

D-Roberts committed:
Fix corresponding return dict bug

D-Roberts committed:
Override test tol

D-Roberts committed:
Change default values of training to False

D-Roberts committed:
Set training to default False X3

D-Roberts committed:
Rm axis from ln

D-Roberts committed:
Set init in dense projection

D-Roberts committed:
Rm debug stuff

D-Roberts committed:
Make style; all tests pass.

D-Roberts committed:
Modify year to 2023

Rocketknight1 commented on src/transformers/models/efficientformer/modeling_tf_efficientformer.py:
```diff
@@ -0,0 +1,1066 @@
+                TFEfficientFormerIntermediateStage(
+                    config,
+                    i,
+                    name=f"intermediate_stages.{2*i}",
+                )
+            )
+            if downsamples[i]:
+                intermediate_stages.append(
+                    TFEfficientFormerPatchEmbeddings(
+                        config, config.hidden_sizes[i], config.hidden_sizes[i + 1], name=f"intermediate_stages.{2*i+1}"
```
Do these names assume every intermediate stage downsamples? It might be safer to keep a counter for the layer names that increments when you append a layer, so that the count is correct no matter which layers do downsampling.

D-Roberts replied:
Oh great catch! Add a layer counter and tested with configs where all downsamples are False and all hidden sizes are the same (then now downsampling layers are added).

Rocketknight1 reviewed (approved):
Overall this looks like an incredibly solid port! I think this might be the best handling of a complex PR like this that I've ever seen from someone not on the Hugging Face payroll. Most issues I raised are just comments or nits, but manipulating `self.ab` during the forward pass and the layer names in the encoder are two that could potentially be breaking. Let me know what you think of the proposed solutions there, but I think this PR should be ready to merge very soon.

Rocketknight1 commented:
cc @amyeroberts for core maintainer review as well

D-Roberts commented on src/transformers/models/efficientformer/modeling_tf_efficientformer.py:
```diff
@@ -0,0 +1,1066 @@
+        return hidden_states
+
+
+class TFEfficientFormerIntermediateStage(tf.keras.layers.Layer):
+    def __init__(self, config: EfficientFormerConfig, index: int, **kwargs):
+        super().__init__(**kwargs)
+        self.meta4D_layers = TFEfficientFormerMeta4DLayers(config=config, stage_idx=index, name="meta4D_layers")
+
+    def call(self, hidden_states: tf.Tensor, training: bool = False) -> Tuple[tf.Tensor]:
+        hidden_states = self.meta4D_layers(hidden_states=hidden_states, training=training)
+        return hidden_states
```
Will leave as is in this iteration then.

amyeroberts commented on src/transformers/models/efficientformer/modeling_tf_efficientformer.py:
```diff
@@ -0,0 +1,1066 @@
+        intermediate_stages = []
+        for i in range(num_intermediate_stages):
+            intermediate_stages.append(
+                TFEfficientFormerIntermediateStage(
+                    config,
+                    i,
+                    name=f"intermediate_stages.{2*i}",
+                )
```
nit: can be on one line 
```suggestion
                TFEfficientFormerIntermediateStage(config, i, name=f"intermediate_stages.{2*i}")
```

amyeroberts reviewed (commented):
Thanks for adding this model! 

Overall a really nice, clean PR, super easy to review 🤗  There's a few places were the architecture implementation deviates for the standard pattern but this seems to come from the PT model. 

In general, just a few comments before we're good to merge: 
* As @Rocketknight1 highlighted, the logic for `self.ab` is very non-canonical and potentially breaking for TF, so let's go for the local var `ab = tf.gather(...)`  logic
* The `serving_output` logic should be updated for the hidden_states and attentions to be conditionally returned based on the config settings, and a comment added about why they're not converted to tensors (different shapes) as in other vision models
* Switching the NHWC format should just happen once in the main layer

D-Roberts committed:
Fix attention biases codes

D-Roberts committed:
Update the shape list logic

D-Roberts commented on src/transformers/models/efficientformer/modeling_tf_efficientformer.py:
```diff
@@ -0,0 +1,1066 @@
+        if training and hasattr(self, "ab"):
+            del self.ab
+        else:
+            self.ab = tf.gather(params=self.attention_biases, indices=self.attention_bias_idxs, axis=1)
```
Thank you for catching this one guys, I agree. I set up an explicit` attention_biases = tf.gather(....)` near where it is being used.

D-Roberts commented on src/transformers/models/efficientformer/modeling_tf_efficientformer.py:
```diff
@@ -0,0 +1,1066 @@
+            x = tf.gather(params=self.attention_biases, indices=self.attention_bias_idxs, axis=1)
+        else:
+            x = self.ab
+        attention_probs = attention_probs + x
```
done.

D-Roberts commented on src/transformers/models/efficientformer/modeling_tf_efficientformer.py:
```diff
@@ -0,0 +1,1066 @@
+            shape_list(hidden_states)[0],
+            shape_list(hidden_states)[1],
+            shape_list(hidden_states)[2],
+        )
```
Nice, done.

D-Roberts committed:
Add a batch norm eps config

D-Roberts commented on src/transformers/models/efficientformer/modeling_tf_efficientformer.py:
```diff
@@ -0,0 +1,1066 @@
+        # Use same default momentum and epsilon as PyTorch equivalent for BatchNormalization
+        self.batchnorm_after = tf.keras.layers.BatchNormalization(
+            axis=-1, epsilon=1e-05, momentum=0.9, name="batchnorm_after"
+        )
```
done

D-Roberts commented on src/transformers/models/efficientformer/modeling_tf_efficientformer.py:
```diff
@@ -0,0 +1,1066 @@
+        # Use same default momentum and epsilon as PyTorch equivalent for BatchNormalization
+        self.batchnorm_after = tf.keras.layers.BatchNormalization(
+            axis=-1, epsilon=1e-05, momentum=0.9, name="batchnorm_after"
+        )
```
done

D-Roberts commented on src/transformers/models/efficientformer/modeling_tf_efficientformer.py:
```diff
@@ -0,0 +1,1066 @@
+        )
+        # Use same default momentum and epsilon as PyTorch equivalent for BatchNormalization
+        self.norm = (
+            tf.keras.layers.BatchNormalization(axis=-1, epsilon=1e-05, momentum=0.9, name="norm")
```
Nice, done.

D-Roberts committed:
Remove extract comments in test files

D-Roberts committed:
Add conditional attn and hidden states return for serving output

D-Roberts commented on src/transformers/models/efficientformer/modeling_tf_efficientformer.py:
```diff
@@ -0,0 +1,1066 @@
+    def serving_output(
+        self, output: TFEfficientFormerForImageClassificationWithTeacherOutput
+    ) -> TFEfficientFormerForImageClassificationWithTeacherOutput:
+        return TFEfficientFormerForImageClassificationWithTeacherOutput(
```
Since the outputs have different dims, I skipped the tensor conversion and added a comment. I had originally followed this example code from [Swin](https://github.com/huggingface/transformers/blob/main/src/transformers/models/swin/modeling_tf_swin.py#L1246), but the conditional return makes sense.

D-Roberts commented on src/transformers/models/efficientformer/modeling_tf_efficientformer.py:
```diff
@@ -0,0 +1,1066 @@
+
+    def serving_output(self, output: TFImageClassifierOutput) -> TFImageClassifierOutput:
+        return TFImageClassifierOutput(
+            logits=output.logits, hidden_states=output.hidden_states, attentions=output.attentions
```
done

D-Roberts commented on src/transformers/models/efficientformer/modeling_tf_efficientformer.py:
```diff
@@ -0,0 +1,1066 @@
+        return TFBaseModelOutput(
+            last_hidden_state=output.last_hidden_state,
+            hidden_states=output.hidden_states,
+            attentions=output.attentions,
```
done

D-Roberts committed:
Change channel dim checking logic

D-Roberts committed:
Add exception for withteacher model in training mode

D-Roberts committed:
Revert layer count for now

D-Roberts commented on src/transformers/models/efficientformer/modeling_tf_efficientformer.py:
```diff
@@ -0,0 +1,1066 @@
+        if tf.executing_eagerly() and num_channels != self.num_channels:
+            raise ValueError(
+                "Make sure that the channel dimension of the pixel values match with the one set in the configuration."
+            )
```
Nice, added an `assert_shapes` solution.

D-Roberts commented on src/transformers/models/efficientformer/modeling_tf_efficientformer.py:
```diff
@@ -0,0 +1,1066 @@
+
+        distillation_logits = self.distillation_classifier(tf.reduce_mean(sequence_output, axis=-2))
+
+        # during inference, return the average of both classifier predictions
```
Removed comment and added an Exception.

D-Roberts committed:
Add layer count for conditional layer naming

D-Roberts committed:
Transpose for conv happens only in main layer

D-Roberts committed:
Make tests smaller

D-Roberts committed:
Make style

D-Roberts commented on tests/models/efficientformer/test_modeling_tf_efficientformer.py:
```diff
@@ -0,0 +1,390 @@
+        num_channels: int = 3,
+        is_training: bool = True,
+        use_labels: bool = True,
+        hidden_size: int = 448,
```
Made smaller in both tf and pt.

D-Roberts commented on src/transformers/models/efficientformer/modeling_tf_efficientformer.py:
```diff
@@ -0,0 +1,1066 @@
+        # When running on CPU, `tf.keras.layers.Conv2D` doesn't support `NCHW` format.
+        # So change the input format from `NCHW` to `NHWC`.
+        # shape = (batch_size, in_height, in_width, in_channels=num_channels)
+        pixel_values = tf.transpose(pixel_values, perm=(0, 2, 3, 1))
```
Nice, this is a lot cleaner.

D-Roberts commented:
@Rocketknight1 @amyeroberts I addressed your comments and also submitted two PRs for the l1 and l3 weights (and tagged Rocketknight1). Let me know what's next!

amyeroberts commented:
@D-Roberts  - that's great! 

For the CI - it seems there is an issue with your CircleCI permissions, as the tests won't run.
Could you try refreshing your permissions as shown [here](https://support.circleci.com/hc/en-us/articles/360048210711-How-to-Refresh-User-Permissions-)? Once all the tests are green, we'll be ready for final reviews :)

D-Roberts committed:
Update doc

D-Roberts commented:
@amyeroberts Thanks for pointing out the circle ci fix. It appears that one doc test which (rightly) can't find tf weights is failing for now. I added back the `from_pt` in the model tests  for the sake of ci tests until we have tf weights to test.

Rocketknight1 commented:
@D-Roberts Just to let you know, we've reached out to the team at Snap to ask them to merge your PRs on the EfficientFormer checkpoints. Sorry for the delay!

Rocketknight1 commented:
@D-Roberts the checkpoint PRs should be merged now. Thank you to @alanspike for the quick response!

D-Roberts committed:
Rm from_pt

D-Roberts commented:
@amyeroberts @Rocketknight1 All local tests pass with the new tf weights. The CI gets this documentation tests failing; the pt version also predicts 281 which maps to label_281 in config.

Rocketknight1 commented:
@D-roberts I think it's fine to swap those tests for just checking the actual argmax index rather than the `id2label` string value. Obviously the repository config doesn't actually have the `id2label` values set, so fixing that would require another PR to the repos.

D-Roberts committed:
Change to actual expect image class label

D-Roberts commented:
@Rocketknight1 @amyeroberts Alright. :)

Rocketknight1 commented:
LGTM now - I'm happy to merge as soon as you and @amyeroberts are!

amyeroberts commented on src/transformers/models/efficientformer/modeling_tf_efficientformer.py:
```diff
@@ -0,0 +1,1042 @@
+        sequence_output = encoder_outputs[0]
+        sequence_output = self.layernorm(sequence_output, training=training)
+
+        # Change the hidden states from NHWC to NCHW; the hidden states are
+        # in NHWC shape after all stages except the MB3D blocks
```
same here re more explicit comment

```suggestion
        # Change the hidden states from (batch_size, height, width, num_channels) to 
        # (batch_size, num_channels, height, width)
        # the hidden states are in (batch_size, height, width, num_channels)
        # shape after all stages except the MB3D blocks
```

amyeroberts reviewed (approved):
Thanks for adding this model! 

All LGTM - just a few tiny nits. Otherwise I think we're good to merge :)

D-Roberts committed:
Remove stray print in tests

D-Roberts committed:
Update image processor test

D-Roberts committed:
Remove the old serving output logic

D-Roberts committed:
Make style

D-Roberts committed:
Make style

D-Roberts committed:
Complete test

D-Roberts commented:
Hi @Rocketknight1 , I've addressed the last comments from @amyeroberts and had all tests pass. I am ready for merge whenever you are. I've just rebased to upstream and there are some unrelated ci tests failing, though last night everything was green.

amyeroberts commented:
@D-Roberts We release [roughly once a month](https://github.com/huggingface/transformers/releases)  and are planning on releasing 4.30 later this week. If you need it right now, it's possible to [install from source](https://huggingface.co/docs/transformers/installation#install-from-source) to have the `main` version too.
